### PR TITLE
Pass the Py_LIMITED_API macro to Extensions

### PR DESCRIPTION
**Pull request by rauletorresc**
**Context:** Extensions also need to be informed about the usage of the limited API.

**Description of the Change:** Pass the `Py_LIMITED_API `macro to Extensions

**Benefits:** Extensions will conform with the limited API.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -306,7 +310,11 @@ def run(self):
 
 ext_modules = [
     custom_calls_extension,
-    CMakeExtension("catalyst.utils.wrapper", sourcedir=frontend_dir),
+    CMakeExtension(
+        "catalyst.utils.wrapper",
+        sourcedir=frontend_dir,
+        define_macros=Py_LIMITED_API_macros,
+    ),
```
[the remark this answers is not in the document]
Comment (dime10) [reply setup.py:317]: Just to confirm, to your knowledge @joeycarter the only module that wasn't already being built with the limited API was the custom calls then?

**Comment by dime10** [comment]
The last thing I'm concerned about is the third-party modules we (atleast attempt) to [ship in our wheels](https://github.com/PennyLaneAI/catalyst/blob/0719a87b4acfd91b178346ee196d9afc1f186594/Makefile#L196-L199):
- CUDA: python only
- OQD: C++ sources, but no Python module (nanobind/pybind) as far as I can see
- OQC: this one has C++ sources and still builds a Python module with 

For OQC the situation is rather confusing, the Python module is built separately from the device backend 
https://github.com/PennyLaneAI/catalyst/blob/0719a87b4acfd91b178346ee196d9afc1f186594/frontend/catalyst/third_party/oqc/src/CMakeLists.txt#L41
https://github.com/PennyLaneAI/catalyst/blob/0719a87b4acfd91b178346ee196d9afc1f186594/frontend/catalyst/third_party/oqc/src/CMakeLists.txt#L55-L59
but we only ship the device backend (without the module):
https://github.com/PennyLaneAI/catalyst/blob/0719a87b4acfd91b178346ee196d9afc1f186594/Makefile#L196-L197

Furthermore, the only thing linking the module to the device backend is the compiler macro `OQC_PY` shown in the cmake snippet above, but since that value is set at build time it must be out of date at install time on a different system! It is used by the device backend here:
https://github.com/PennyLaneAI/catalyst/blob/0719a87b4acfd91b178346ee196d9afc1f186594/frontend/catalyst/third_party/oqc/src/OQCRunner.hpp#L114

To recap, the OQC plugin is still being built with Pybind11, which is not stable abi compatible, but the relevant module isn't even shipped in the wheels or configured for distribution 🤔

File: setup.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rauletorresc) [reply setup.py]: Removed

[the remark this answers is not in the document]
Comment (joeycarter) [reply setup.py]: @dime10, yes, as far as I know we do not need to explicitly opt in to the limited API anywhere else. In fact, it might not even be strictly necessary for custom_calls, since it doesn't actually use the Python C API at the moment, but it makes our intention explicit should this change in the future.

**Comment by joeycarter** [comment]
> The last thing I'm concerned about is the third-party modules we (atleast attempt) to [ship in our wheels](https://github.com/PennyLaneAI/catalyst/blob/0719a87b4acfd91b178346ee196d9afc1f186594/Makefile#L196-L199):
>
> * CUDA: python only
> * OQD: C++ sources, but no Python module (nanobind/pybind) as far as I can see
> * OQC: this one has C++ sources _and_ still builds a Python module with Pybind11

The OQ**D** modules shouldn't cause issues with building the wheels with the stable ABI since they don't use the Python C API at all (for now, at least).

For OQ**C**, I think you're right that this is not correct. I just tried building the wheels locally and running a simple test and it fails:

```python
import pennylane as qml
from catalyst.third_party.oqc import OQCDevice

import os
os.environ["OQC_EMAIL"] = "email"
os.environ["OQC_PASSWORD"] = "password"
os.environ["OQC_URL"] = "oqc_url"

device = OQCDevice(backend="lucy", shots=1000, wires=1)
@qml.qjit
@qml.qnode(device)
def counts_1qbit(x: float):
    qml.RX(x, wires=0)
    return qml.counts()

counts_1qbit(1.0)
```

```
RuntimeError: [.../catalyst/frontend/catalyst/third_party/oqc/src/../../../../../runtime/include/DynamicLibraryLoader.hpp][Line:75][Function:getSymbol] Error in Catalyst Runtime: .../catalyst/frontend/catalyst/third_party/oqc/src/build/oqc_python_module.cpython-312-x86_64-linux-gnu.so: undefined symbol: counts
```

I think the reason we missed this is because `oqc_python_module` only defines a `counts()` function, but as far as I can tell we have no tests covering it.

I don't think the OQC issue is blocking this PR, but we should address it in a future PR.

**Comment by joeycarter** [comment]
Furthermore, we skip the OQC frontend tests entirely if the `qcaas_client` package is not installed:

https://github.com/PennyLaneAI/catalyst/blob/0719a87b4acfd91b178346ee196d9afc1f186594/frontend/test/test_oqc/conftest.py#L22-L27

https://github.com/PennyLaneAI/catalyst/blob/0719a87b4acfd91b178346ee196d9afc1f186594/frontend/test/test_oqc/conftest.py#L44-L49

**Review by joeycarter — APPROVED** [review]
Happy to approve, but we should look into the OQC issue I mentioned above.

**Comment by dime10** [comment]
> I don't think the OQC issue is blocking this PR, but we should address it in a future PR.

So I think it's relevant to this PR in the sense that to properly ship OQC with the stable abi we need to convert it to nanobind, but you're right we can do this in a separate PR 👍

**Comment by rauletorresc** [comment]
> > I don't think the OQC issue is blocking this PR, but we should address it in a future PR.
> 
> So I think it's relevant to this PR in the sense that to properly ship OQC with the stable abi we need to convert it to nanobind, but you're right we can do this in a separate PR 👍

https://github.com/PennyLaneAI/catalyst/pull/1391
